storage/include/sparse_merkle/internal_node.h

```diff
@@ -0,0 +1,280 @@
+// Concretely, if the displayed BatchedInternalNode is at depth 0 (height 256)
+// of the entire sparse merkle tree, then 2 keys with 5 leading bits in common
+// would cause `C` to be created and put at height 0 of this
+// BatchedInternalNode. This scenario is demonstrated in the test
+// `split_until_new_batch_node_needed` in `internal_node_test.cpp`.
+//
+*/
+class BatchedInternalNode {
+ public:
+  static constexpr size_t MAX_CHILDREN = 31;
+
+  // The leaf was inserted into this node successfully.
+  //
+  // It's possible that a leaf with the same key hash as the newly inserted leaf
+  // was overwritten.  In this case we need to tell the caller that this
+  // occurred so they can add it to their stale list for pruning later.
+  struct InsertComplete {
+    std::optional<LeafKey> stale_leaf;
+  };
+
+  // A node split has occurred, forcing the need to create new
+  // BatchedInternalNodes.
+  //
+  // Depending upon how many bits in common the key hashes of these 2 leaves
+  // have, more than one BatchedInternalNode may need to be created.
+  //
+  // Return the original child that forced a node split.
+  struct CreateNewBatchedInternalNodes {
+    // This LeafChild was previously stored in this BatchedInternalNode, but was replaced
+    // with an InternalChild. It now needs to be inserted in a new node along
+    // with the LeafChild originally trying to be inserted.
+    LeafChild stored_child;
+  };
+
+  // A leaf could not be added at height 0 because there was an existing
+  // InternalChild at the leaf (height 0) of this BatchedInternalNode.
+  //
+  // An attempt should be made to insert the leaf at the next, already existing
+  // BatchedInternalNode with `next_node_version`.
+  struct InsertIntoExistingNode {
+    // The version of the next BatchedInternalNode to try inserting into. The
+    // caller knows how to construct an InternalNodeKey using this version.
+    Version next_node_version;
+  };
+
+  // insert(child, depth) can return one of three results to the sparse merkle
+  // tree manipulating the BatchedInternalNode.
+  typedef std::variant<InsertComplete, CreateNewBatchedInternalNodes, InsertIntoExistingNode> InsertResult;
+
+  // Insert a LeafChild into this internal node. Return a type indicating success or
+  // failure that contains the necessary data for the caller (the sparse merkle
+  // tree implementation) to do the right thing.
+  //
+  // `depth` represents how many nibbles down the sparse merkle tree this BatchedInternalNode is.
+  InsertResult insert(LeafChild& child, size_t depth);
+
+  // Return the root hash of this node.
+  const Hash& hash() const { return get_hash(0); }
+
+  // Return the version of the root node of this BatchedInternalNode
+  Version version() const {
+    auto& root = children_[0];
+    if (!root) return 0;
+    return std::get<InternalChild>(root.value()).version;
+  }
+
+  // Return the number of childrent that are not std::nullopt.
+  size_t num_children() const {
```

Comment:
Most of our code has mixed upper-low case function name conventions, like 'executeNextCommittedRequests'.

Reply:
Good catch. I was indeed not following the style guide. Fixed.

Now I get to point out when you all use camelCase variable names 😆